### PR TITLE
Use Tink for OIDC Keysets, with rotation

### PR DESCRIPTION
**Pull request by lstoll**
Adds a keyset manager, to manage rotating a tink JWT keyset in the DB.

Pulls in a just new enough version of lstoll/oidc to support this.

Just using raw keys in the file - I don't think this app/the way it runs makes it worth forcing a KMS, for now anyway.

### CODE REVIEW

File: keyset.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+		Template:    jwt.RS256_2048_F4_Key_Template(),
+		RotateEvery: 24 * time.Hour,
+	}
+	KeysetCookie = Keyset{
+		Name:        "cookie",
+		Template:    aead.AES128GCMSIVKeyTemplate(),
+		RotateEvery: 30 * 24 * time.Hour,
+	}
+
+	allKeysets = []Keyset{KeysetOIDC, KeysetCookie}
```
Comment (sr) [inline keyset.go:40]: it's beautiful 😢

File: db.go [hunk 2]
```diff
@@ -345,45 +317,26 @@ func (db *DB) GetAuthenticatedUser(sessionID string) (AuthenticatedUser, error)
...
-	}
-	return &staticKeySource{key: v}
-}
-
 func (db *DB) SessionManager() core.SessionManager {
 	return &sessionManager{f: db.f}
 }
 
-const rsaKeyBits = 2048
+func (db *DB) GetKeyset(ks Keyset) DBKeyset {
```
Comment (sr) [inline db.go:324]: nit but I would change this to just take a `name string` (ditto for `PutKeySet`)

File: keyset.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+func (k *KeysetHandles) Handle(context.Context) (*keyset.Handle, error) {
+	rdr := keyset.NewJSONReader(bytes.NewReader(k.db.GetKeyset(k.keyset).Keyset))
+	h, err := insecurecleartextkeyset.Read(rdr)
+	if err != nil {
+		return nil, fmt.Errorf("parsing keyset from db: %w", err)
+	}
+	return h, nil
+}
+
+// PublicHandle returns a handle to the current keyset, only including private
```
Comment (sr) [inline keyset.go:60]:
```suggestion
// PublicHandle returns a handle to the current keyset, only including public
```

File: keyset.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+// PrivateHandle returns a handle to the current keyset, including private keys
+func (o *KeysetManager) Handles(keyset Keyset) *KeysetHandles {
+	return &KeysetHandles{db: o.db, keyset: keyset}
+}
+
+// Run synchronously runs a loop to rotate the keys in the DB as needed, and
+// update handles on this manager instance.
+func (o KeysetManager) Run() error {
+	t := time.NewTicker(keysetRotateCheckInterval)
```
Comment (sr) [inline keyset.go:101]:
```suggestion
	t := time.NewTicker(keysetRotateCheckInterval)
	defer t.Stop()
```

File: main.go [hunk 5]
```diff
@@ -119,14 +120,23 @@ func main() {
 	}
 }
 
-func serve(ctx context.Context, db *DB, keyset *derivedKeyset, issuer issuerConfig, _, _ time.Duration, addr string) error {
+func serve(ctx context.Context, db *DB, appKeys *derivedKeyset, issuer issuerConfig, _, _ time.Duration, addr string) error {
```
Comment (sr) [inline main.go:123]: Can the two unused time.Duration params be deleted now?

File: main.go [hunk 6]
```diff
@@ -280,7 +291,9 @@ func serve(ctx context.Context, db *DB, keyset *derivedKeyset, issuer issuerConf
 
 	var g run.Group
 
-	g.Add(run.SignalHandler(ctx, os.Interrupt))
+	g.Add(run.SignalHandler(ctx, os.Interrupt, unix.SIGTERM))
```
Comment (sr) [inline main.go:294]: 😬

**Review by sr — APPROVED** [review]
love it

File: main.go (hunk 5 above)
Comment (lstoll) [reply main.go]: removed

**Pull request merged**